### PR TITLE
Create Todo List

### PR DESCRIPTION
**Pull request by mustaqeemtariq**
This PR contains the implementation of CRUD operations using web api. 

Task is attached for reference:
https://dev.azure.com/CodeNinjaConsulting/Talent%20Pool/_workitems/edit/1411

### CODE REVIEW

File: Models/TodoContext.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+using TodoApi.Models;
+
+    public class TodoContext : DbContext
+    {
+        public TodoContext (DbContextOptions<TodoContext> options)
+            : base(options)
+        {
+        }
+
+        public DbSet<TodoApi.Models.TodoItem> TodoItem { get; set; } = default!;
```
Comment (taimoor0nline) [inline Models/TodoContext.cs]: TodoApi.Models provide this reference on top of the class as name space and then type should be defined

File: Program.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
+using System.Configuration;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Configuration;
+using Microsoft.EntityFrameworkCore.InMemory;
```
Comment (taimoor0nline) [inline Program.cs]: can you please explain the purpose of using InMemory here ?

File: Program.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
+using System.Configuration;
```
Comment (taimoor0nline) [inline Program.cs]: if system. configuration is not being used remove it.

File: Program.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+if (app.Environment.IsDevelopment())
+{
+    app.UseSwagger();
+    app.UseSwaggerUI();
+}
+
+app.UseHttpsRedirection();
+
+app.UseAuthorization();
```
Comment (taimoor0nline) [inline Program.cs]: are you using app.UseAuthorization() if yes where is UseAuthentication() ??

File: Controllers/TodoItemsController.cs
Comment (taimoor0nline) [inline Controllers/TodoItemsController.cs:1]: Please use try catch to handle exception , you may use global level exception handling via midlewear

**Review by taimoor0nline — CHANGES_REQUESTED** [review]
Please review the changes

**Review by taimoor0nline — CHANGES_REQUESTED** [review]
Please review changes

File: Program.cs (hunk 2 above)
Comment (mustaqeemtariq) [reply Program.cs]: First I was running the code without database so using local memory for saving data so it was being used for that purpose.

File: Program.cs (hunk 3 above)
Comment (mustaqeemtariq) [reply Program.cs]: Done

File: Program.cs (hunk 4 above)
Comment (mustaqeemtariq) [reply Program.cs]: No it is not being used.

**Comment by mustaqeemtariq** [comment]
@taimoor0nline I have updated the code. Kindly review now. Thanks

File: Models/TodoContext.cs (hunk 1 above)
Comment (mustaqeemtariq) [reply Models/TodoContext.cs]: Done

File: Controllers/TodoItemsController.cs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+        public TodoItemsController(TodoContext context)
+        {
+            _context = context;
+        }
+
+        [HttpGet]
+        public async Task<ActionResult<IEnumerable<TodoItem>>> GetTodoItem()
+        {
+            if (_context.TodoItem == null)
```
Comment (taimoor0nline) [inline Controllers/TodoItemsController.cs]: _context.TodoItem == will not work please fix it

File: Controllers/TodoItemsController.cs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+            {
+                return NotFound();
+            }
+            return await _context.TodoItem.ToListAsync();
+        }
+
+        [HttpGet("{id}")]
+        public async Task<ActionResult<TodoItem>> GetTodoItem(long id)
+        {
+            if (_context.TodoItem == null)
```
Comment (taimoor0nline) [inline Controllers/TodoItemsController.cs]: same as above

File: Controllers/TodoItemsController.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+        [HttpPut("{id}")]
+        public async Task<IActionResult> PutTodoItem(long id, TodoItem todoItem)
+        {
+            if (id != todoItem.Id)
+            {
+                return BadRequest();
+            }
+
+            _context.Entry(todoItem).State = EntityState.Modified;
```
Comment (taimoor0nline) [inline Controllers/TodoItemsController.cs:56]: Please validate your model before performing database operation

File: Controllers/TodoItemsController.cs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+            }
+
+            return NoContent();
+        }
+
+        [HttpPost]
+
+        public async Task<ActionResult<TodoItem>> PostTodoItem(TodoItem todoItem)
+        {
+            _context.TodoItem.Add(todoItem);
```
Comment (taimoor0nline) [inline Controllers/TodoItemsController.cs:90]: please check for null and model validity

File: Controllers/TodoItemsController.cs [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+            await _context.SaveChangesAsync();
+
+            return CreatedAtAction(nameof(GetTodoItem), new { id = todoItem.Id }, todoItem);
+        }
+
+        [HttpDelete("{id}")]
+        public async Task<IActionResult> DeleteTodoItem(long id)
+        {
+            if (_context.TodoItem == null)
```
Comment (taimoor0nline) [inline Controllers/TodoItemsController.cs]: this condition is wrongly written , how you can check if todo item is null aslo for naming convention please follow the guide line given in below link https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/identifier-names

**Review by taimoor0nline — CHANGES_REQUESTED** [review]
Please use async await where applicable

File: Controllers/TodoItemsController.cs (hunk 5 above)
Comment (mustaqeemtariq) [reply Controllers/TodoItemsController.cs]: Fixed

File: Controllers/TodoItemsController.cs (hunk 6 above)
Comment (mustaqeemtariq) [reply Controllers/TodoItemsController.cs]: Fixed

File: Controllers/TodoItemsController.cs (hunk 7 above)
Comment (mustaqeemtariq) [reply Controllers/TodoItemsController.cs:56]: Done

File: Controllers/TodoItemsController.cs (hunk 8 above)
Comment (mustaqeemtariq) [reply Controllers/TodoItemsController.cs:90]: Done

File: Controllers/TodoItemsController.cs (hunk 9 above)
Comment (mustaqeemtariq) [reply Controllers/TodoItemsController.cs]: Done

**Review by taimoor0nline — APPROVED** [review]

**Pull request merged**